### PR TITLE
Allow services to use System.Text.Json for OOP comms in Razor

### PR DESCRIPTION
**Pull request by davidwengier**
Given the amount of inheritance and variation in the LSP types, between VS and VS Code etc. the original plan to wrap LSP types for OOP comms, or add MessagePack compatible attributes, or something, seems way too hard. This is much easier :)

The use of Json is opt-in per service, so this will let us pick and choose the best method on a case-by-case basis.

### CODE REVIEW

**Comment by davidwengier** [comment]
@dibarbet @tmat @CyrusNajmabadi PTAL. Mostly Razor EA changes, but one real roslyn change in the remote infra

**Comment by davidwengier** [comment]
Ping @dotnet/roslyn-ide for reviews please

File: src/Tools/ExternalAccess/Razor/Remote/JsonSerializableRazorPinnedSolutionInfoWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+using System.Text.Json.Serialization;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    /// <summary>
+    /// A wrapper for a solution that can be used by Razor for OOP services that communicate via System.Text.Json
+    /// </summary>
+    internal readonly record struct JsonSerializableRazorPinnedSolutionInfoWrapper(
+        [property: JsonPropertyName("data1")] long Data1,
+        [property: JsonPropertyName("data2")] long Data2)
```
Comment (CyrusNajmabadi) [inline src/Tools/ExternalAccess/Razor/Remote/JsonSerializableRazorPinnedSolutionInfoWrapper.cs:15]: interesting.  would like to know more about this.  when are you guuys pinning solutions?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Tools/ExternalAccess/Razor/Remote/JsonSerializableRazorPinnedSolutionInfoWrapper.cs (hunk 1 above)
Comment (davidwengier) [reply src/Tools/ExternalAccess/Razor/Remote/JsonSerializableRazorPinnedSolutionInfoWrapper.cs:15]: > interesting. would like to know more about this. when are you guuys pinning solutions?

hopefully never? This is just a version of [RazorPinnedSolutionInfoWrapper](https://github.com/davidwengier/roslyn/blob/06d3f153ed6af6f2b78028a1e1e6ecc55c8ff101/src/Tools/ExternalAccess/Razor/Remote/RazorPinnedSolutionInfoWrapper.cs#L11), which is really just [Checksum](https://github.com/davidwengier/roslyn/blob/main/src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs#L24), but this serializes with STJ and that uses messagepack. I could add STJ attributes to `RazorPinnedSolutionInfoWrapper`, but then I'd need them on `Checksum` too.

Comment (eiriktsarpalis) [reply src/Tools/ExternalAccess/Razor/Remote/JsonSerializableRazorPinnedSolutionInfoWrapper.cs:15]: > @eiriktsarpalis how do longs get serialized in STJ? Are they guaranteed to roundtrip?

They serialize to JSON numbers which have arbitrary precision. They are guaranteed to roundtrip.

Comment (CyrusNajmabadi) [reply src/Tools/ExternalAccess/Razor/Remote/JsonSerializableRazorPinnedSolutionInfoWrapper.cs:15]: Thanks!

**Pull request merged**